test/e2e/federated-service.go

```diff
@@ -0,0 +1,209 @@
+						{
+							ClientCIDR:    "0.0.0.0/0",
+							ServerAddress: context.Cluster.Cluster.Server,
+						},
+					},
+				},
+			}
+			_, err := f.FederationClientset.Federation().Clusters().Create(&cluster)
+			framework.ExpectNoError(err, "Creating cluster")
+		}
+
+		By("Obtaining a list of all the clusters")
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		framework.Logf("%d clusters found", len(clusterList.Items))
+
+		for _, cluster := range clusterList.Items {
+			framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+
+			Expect(framework.TestContext.KubeConfig).ToNot(Equal(""), "KubeConfig must be specified to load clusters' client config")
+			kubecfg, err := clientcmd.LoadFromFile(framework.TestContext.KubeConfig)
+			framework.ExpectNoError(err, "error loading KubeConfig: %v", err)
+
+			cfgOverride := &clientcmd.ConfigOverrides{
+				ClusterInfo: clientcmdapi.Cluster{
+					Server: cluster.Spec.ServerAddressByClientCIDRs[0].ServerAddress,
+				},
+			}
+			ccfg := clientcmd.NewNonInteractiveClientConfig(*kubecfg, cluster.Name, cfgOverride, clientcmd.NewDefaultClientConfigLoadingRules())
+			cfg, err := ccfg.ClientConfig()
+			Expect(err).NotTo(HaveOccurred())
+
+			cfg.QPS = KubeAPIQPS
+			cfg.Burst = KubeAPIBurst
+			clset := release_1_3.NewForConfigOrDie(restclient.AddUserAgent(cfg, UserAgentName))
+			clusterClientSets = append(clusterClientSets, clset)
+		}
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		discoverService(f)
+	})
+
+	It("should be able to discover a non-local federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		// Delete a federated service shard in the default e2e Kubernetes cluster.
+		err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+
+		discoverService(f)
+	})
+})
+
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
+	By("Fetching a federated service shard")
+	var clSvcList *v1.ServiceList
+	for t := time.Now(); time.Since(t) < FederatedServiceTimeout; time.Sleep(framework.Poll) {
+		var err error
+		clSvcList, err = cs.Core().Services(namespace).List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		n := len(clSvcList.Items)
+		if n == numSvcs {
+			break
+		}
+		framework.Logf("%d services found, trying again in %s", n, framework.Poll)
+	}
+
+	Expect(len(clSvcList.Items)).To(Equal(numSvcs))
+	if numSvcs > 0 {
+		// Renaming for clarity/readability
+		clSvc := clSvcList.Items[0]
```

Comment:
This only seems to work for numSvcs == 1 ?  The semantics of this function are rather confusing.  As far as I can tell it will wait until the number of services in the specifed namespace reaches numSvcs, and (presumably this only works if 'service' is not nil) if numSvcs >0, that the first listed service matches the one passed in?  You should definitely document the semantics in the header, and also check for service != nil before dereferencing it.